app/controllers/MrnEntryController.scala

```diff
@@ -50,36 +51,56 @@ class MrnEntryController @Inject()(
 
   val form = formProvider()
 
-  lazy val backLinkUrl: String =
+  lazy val defautBackLinkUrl: String =
     if (secureMessagingConfig.isSecureMessagingEnabled) routes.ChoiceController.onPageLoad.url
     else routes.StartController.displayStartPage.url
 
-  def onPageLoad: Action[AnyContent] = (authenticate andThen verifiedEmail andThen getData) { implicit req =>
+  private def getBackLink(refererUrl: Option[String]) = refererUrl.getOrElse(defautBackLinkUrl)
+
+  def onPageLoad(refererUrl: Option[String] = None): Action[AnyContent] = (authenticate andThen verifiedEmail andThen getData).async { implicit req =>
     val populatedForm = req.userAnswers.mrn.fold(form)(form.fill)
-    Ok(mrnEntry(populatedForm, backLinkUrl))
+
+    decodeRefererUrl(refererUrl).map { decodedBackLink =>
+      answersService
+        .upsert(req.userAnswers.copy(mrnPageRefererUrl = Some(decodedBackLink)))
+        .flatMap(_ => Future.successful(Ok(mrnEntry(populatedForm, getBackLink(Some(decodedBackLink))))))
```

Comment:
Nit: You could remove `Future.successful` from this line if you'd only change `flatMap` to `map`.

Reply:
:+1:

Reply:
And probably ust my personal taste, but after inserting the url into cache I'd merge this case with the default one. It's just that it took me a while to figure out what's the difference between `decodedBackLink` and `getBackLink(...)`